### DISCUSSION TITLE
Does refit support this usage?

### DISCUSSION THREAD

**Discussion by LimOpsx** [reactiveui/refit#1880] [Q&A] [answered]
`[Url("http://localhost:10000/api/user")]
public interface IUserClient
{
    
    [Get("/getUser")]
    public Task<User> GetUser(string username,string passwordHash);
    
}`,
OR 
IUserClient.CS
`[Url("userServerAddress")]
public interface IUserClient
{
    
    [Get("/getUser")]
    public Task<User> GetUser(string username,string passwordHash);
    
}` Then read the request address from appsettings. json
"userServerAddress":"http://localhost:10000/api/user"

**Comment by TimothyMakkison** [comment]
I'm afraid Refit doesn't support either examples. In your case i would do the following:

```C#
var client = RestService.For<IUserClient>("http://localhost:10000/api/user");

var user = await client.GetUser("LimOpsx", "myPassword");
// calls http://localhost:10000/api/user/getUser?username=LimOpsx&passwordHash=myPassword

public interface IUserClient
{
    [Get("/getUser")]
    public Task<User> GetUser(string username,string passwordHash); 
}
```

**Reply by LimOpsx** [reply]
Thanks😊, I've seen this in the official documentation, but I think it's hard coded and it's my first time trying to use a similar library. If I have many Client interfaces, I can only specify many baseaddress relationships.
I tried to extend a feature myself, marking the corresponding Client interfaces, and then obtaining these interfaces through reflection, into the baseaddress configuration.
However, I don't know if this way is the best solution, and I would like to ask you how to solve this scenario in actual development.

**Comment by glennawatson** [comment, marked as answer]
As Timothy said, neither form is supported - there is no `[Url]` attribute in Refit, and the base address is not read from interface metadata or appsettings automatically. The base address is supplied when the client is created.

For your "many clients, base addresses from appsettings.json" scenario, use `IHttpClientFactory` + DI. That is the idiomatic way and avoids hand-rolled reflection.

appsettings.json:

```json
{
  "Services": {
    "UserApi": "http://localhost:10000/api/user",
    "OrderApi": "http://localhost:10001/api/order"
  }
}
```

Registration:

```csharp
services.AddRefitClient<IUserClient>()
    .ConfigureHttpClient((sp, c) =>
        c.BaseAddress = new Uri(
            sp.GetRequiredService<IConfiguration>()["Services:UserApi"]!));

services.AddRefitClient<IOrderClient>()
    .ConfigureHttpClient((sp, c) =>
        c.BaseAddress = new Uri(
            sp.GetRequiredService<IConfiguration>()["Services:OrderApi"]!));
```

Then just inject `IUserClient` / `IOrderClient` where you need them. If you have many clients you can wrap that registration in a small extension method that takes the config key, so each line stays a one-liner. No reflection or custom attribute needed.

If you are not using DI, pass the address explicitly: `RestService.For<IUserClient>(config["Services:UserApi"])`.
